net/libfab/libfab.c

```diff
@@ -0,0 +1,2829 @@
+		if (socktype[s] != NULL) {
+			shift = strlen(socktype[s]);
+			if (strncmp(ep_name, socktype[s], shift) == 0)
+				break;
+		}
+	}
+	if (ep_name[shift] != ':')
+		return M0_ERR(-EINVAL);
+	ep_name += shift + 1;
+	at = strchr(ep_name, '@');
+	if (at == NULL) {
+		return M0_ERR(-EINVAL);
+	} else {
+		at++;
+		if (at == NULL)
+			return M0_ERR(-EINVAL);
+		M0_PRE(portSize >= (strlen(at)+1));
+		memcpy(port,at,(strlen(at)+1));
+	}
+	M0_PRE(nodeSize >= (at - ep_name));
+	memcpy(node, ep_name, ((at - ep_name)-1));
+	return 0;
+}
+
+/**
+ * Used to decode the ip and port from the given end point
+ * ep_name : endpoint address from domain
+ * node    : copy ip address from ep_name
+ * port    : copy port number from ep_name
+ * Example of ep_name IPV4 192.168.0.1:4235
+ *                    IPV6 [4002:db1::1]:4235
+ */
+static int libfab_ep_addr_decode_native(const char *ep_name, char *node,
+					size_t nodeSize, char *port,
+					size_t portSize)
+{
+	char   *name;
+	char   *cp;
+	size_t  n;
+	int     rc = 0;
+
+	M0_PRE(ep_name != NULL);
+
+	M0_ENTRY("ep_name=%s", ep_name);
+
+	name = (char *)ep_name + strlen("libfab:");
+
+	if( name[0] == '[' ) {
+		/* IPV6 pattern */
+		cp = strchr(name, ']');
+		if (cp == NULL)
+			return M0_ERR(-EINVAL);
+
+		name++;
+		n = cp - name;
+		if (n == 0 )
+			return M0_ERR(-EINVAL);
+		cp++;
+		if (*cp != ':')
+		return M0_ERR(-EINVAL);
+		cp++;
+	}
+	else {
+		/* IPV4 pattern */
+		cp = strchr(name, ':');
+		if (cp == NULL)
+			return M0_ERR(-EINVAL);
+
+		n = cp - name;
+		if (n == 0 )
+			return M0_ERR(-EINVAL);
+
+		++cp;
+	}
+
+	M0_PRE(nodeSize >= (n+1));
+	M0_PRE(portSize >= (strlen(cp)+1));
+
+	memcpy(node, name, n);
+	node[n] = 0;
+
+	n=strlen(cp);
+	memcpy(port, cp, n);
+	port[n] = 0;
+
+	return M0_RC(rc);
+}
+
+/**
+ * Parses network address.
+ *
+ * The following address formats are supported:
+ *
+ *     - lnet compatible, see nlx_core_ep_addr_decode():
+ *
+ *           nid:pid:portal:tmid
+ *
+ *       for example: "10.0.2.15@tcp:12345:34:123" or
+ *       "192.168.96.128@tcp1:12345:31:*"
+ *
+ *     - sock format, see socket(2):
+ *           family:type:ipaddr[@port]
+ *
+ *     - libfab compatible format
+ *       for example IPV4 libfab:192.168.0.1:4235
+ *                   IPV6 libfab:[4002:db1::1]:4235
+ *
+ */
+static int libfab_ep_addr_decode(struct m0_fab__ep *ep, const char *name)
+{
+	char *node = ep->fep_name.fen_addr;
+	char *port = ep->fep_name.fen_port;
+	size_t nodeSize = ARRAY_SIZE(ep->fep_name.fen_addr);
+	size_t portSize = ARRAY_SIZE(ep->fep_name.fen_port);
+	int result;
+
+	M0_ENTRY("name=%s", name);
+	
+	if( name == NULL || name[0] == 0)
+		result =  M0_ERR(-EPROTO);
+	else if((strncmp(name,"libfab",6))==0)
+		result = libfab_ep_addr_decode_native(name, node, nodeSize, 
+						      port, portSize);
+	else if (name[0] < '0' || name[0] > '9')
+		/* sock format */
+		result = libfab_ep_addr_decode_sock(name, node, nodeSize, 
+						    port, portSize);
```

Comment:
I think the sock format starts with "unix:" or "inet:" or "inet6:".
Is it?

Reply:
Yes